service/operator/service.go

```diff
@@ -43,14 +121,161 @@ func New(config Config) (*Service, error) {
 // Service implements the service.
 type Service struct {
 	// Dependencies.
-	logger micrologger.Logger
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
 
 	// Internals.
-	bootOnce sync.Once
+	ipsPorts  map[string]int
+	bootOnce  sync.Once
+	mutex     sync.Mutex
+	namespace string
+	service   string
+	tpr       *tpr.TPR
 }
 
+// Boot starts the service and implements the watch for the cluster TPR.
 func (s *Service) Boot() {
 	s.bootOnce.Do(func() {
-		s.logger.Log("debug", "not implemented")
+		err := s.tpr.CreateAndWait()
+		if tpr.IsAlreadyExists(err) {
+			s.logger.Log("debug", "third party resource already exists")
+		} else if err != nil {
+			s.logger.Log("error", fmt.Sprintf("%#v", err))
+			return
+		}
+
+		s.logger.Log("debug", "starting list/watch")
+
+		newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.addFunc,
+			DeleteFunc: s.deleteFunc,
+		}
+		newZeroObjectFactory := &tpr.ZeroObjectFactoryFuncs{
+			NewObjectFunc:     func() runtime.Object { return &kvmtpr.CustomObject{} },
+			NewObjectListFunc: func() runtime.Object { return &kvmtpr.List{} },
+		}
+
+		s.tpr.NewInformer(newResourceEventHandler, newZeroObjectFactory).Run(nil)
 	})
 }
+
+func (s *Service) addFunc(obj interface{}) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	err := s.addFuncError(obj)
+	if err != nil {
+		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+	}
+}
+
+func (s *Service) addFuncError(obj interface{}) error {
+	customObject, ok := obj.(*kvmtpr.CustomObject)
+	if !ok {
+		return microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", &kvmtpr.CustomObject{}, obj)
+	}
+
+	var err error
+
+	err = s.addPortToService(*customObject)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	return nil
+}
+
+// addPortToService adds a cluster port to the Kubernetes service resource of
+// the configured ingress controller, if it does not already exists.
+func (s *Service) addPortToService(customObject kvmtpr.CustomObject) error {
+	k8sService, err := s.k8sClient.CoreV1().Services(s.namespace).Get(s.service)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	portName := PortName(customObject)
+
+	exists := portNameExists(k8sService.Spec.Ports, portName)
+	if exists {
+		s.logger.Log("debug", fmt.Sprintf("port for cluster '%s' already exists", ClusterID(customObject)))
+		return nil
+	}
+
+	clusterPort, err := s.newClusterPort(k8sService.Spec.Ports, ipsPortsToPorts(s.ipsPorts))
```

Comment:
I see `ipsPortsToPorts(s.ipsPorts) == availabePorts` Again I don't see where it's maintained. Am I missing something?